### PR TITLE
Improved parsing tokens

### PR DESCRIPTION
**Pull request by melex750**
* Parsing time is reduced by half.
* Improved unicode support.

### CODE REVIEW

File: CBot/src/CBot/CBotUtils.h [hunk 1]
```diff
@@ -180,4 +187,19 @@ template<typename T> class CBotDoublyLinkedList
...
+{
+    static const uint32_t Table[256];
+};
+
+static inline void InitCRC32(uint32_t& check)
+{
+    check = 0xFFFFFFFF;
+}
+
+static inline void UpdateCRC32(const unsigned char c, uint32_t& check)
```
Comment (hexagonrecursion) [inline CBot/src/CBot/CBotUtils.h:200]:
```suggestion
static inline void UpdateCRC32(unsigned char c, uint32_t& check)
```

Comment (melex750) [reply CBot/src/CBot/CBotUtils.h:200]: Its left over from when I tried passing a reference there.

**Review by hexagonrecursion — CHANGES_REQUESTED** [review]

File: CBot/src/CBot/CBotUtils.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (melex750) [reply CBot/src/CBot/CBotUtils.h:200]: Oops, another left over keyword. 😬

File: CBot/src/CBot/CBotToken.h [hunk 2]
```diff
@@ -251,4 +283,16 @@ extern bool IsOfTypeList(CBotToken* &p, int type1, ...);
...
 const std::string& UndefinedTokenString();
 
+/**
+ * \brief Class containing data about a compiled list of tokens.
+ *
+ */
+struct CBotToken::Data
+{
+    //! CRC32 signature for the list of tokens.
+    uint32_t        signature{0};
```
[the remark this answers is not in the document]
Comment (melex750) [reply CBot/src/CBot/CBotToken.h:293]: ```CBotProgram::RestoreState()``` already starts the program from the beginning if restoring the stack fails. We can do this also if we detect that the cbot code has changed between save and restore.

File: CBot/src/CBot/CBotToken.cpp [hunk 3]
```diff
@@ -241,219 +241,240 @@ void CBotToken::SetPos(int start, int end)
...
+
+CBotToken* CBotToken::NextToken(const char*& program, CBotToken::Data* tokendata, bool first)
 {
-    std::string token; // found token
-    std::string sep;   // separators after the token
-    bool stop = first;
 
     if (*program == 0) return nullptr;
 
     char c = *(program++);                 // next character
```
[the remark this answers is not in the document]
Comment (melex750) [reply CBot/src/CBot/CBotToken.cpp:276]: next octet of booleans :laughing:

File: CBot/src/CBot/CBotToken.cpp [hunk 4]
```diff
@@ -241,219 +241,240 @@ void CBotToken::SetPos(int start, int end)
...
+{
+    return ('\040' < c && c < '\060') || ('\071' < c && c < '\100') ||
+           ('\132' < c && c < '\137') || ('\172' < c && c < '\177');
+}
 
-////////////////////////////////////////////////////////////////////////////////
-CBotToken*  CBotToken::NextToken(const char*& program, bool first)
+inline bool OmitFromStrings(const char c)
+{
+    return c == 0 || c == '\t' || c == '\n' || c == '\r';
```
[the remark this answers is not in the document]
Comment (melex750) [reply CBot/src/CBot/CBotToken.cpp:259]: That's how they were before, i didn't change it.  Maybe it's a good thing; ```\t``` might easier recognize as a tab when you're actually looking at the text.

File: CBot/src/CBot/CBotToken.cpp [hunk 5]
```diff
@@ -241,219 +241,240 @@ void CBotToken::SetPos(int start, int end)
...
+        }
+        //tokenLength ++;
+        tokendata->currentColumnIndex ++;
+    };
 
+    if (!first)
+    {
+        NextCharacter();
         // special case for strings
         if (token[0] == '\"' )
```
[the remark this answers is not in the document]
Comment (melex750) [reply CBot/src/CBot/CBotToken.cpp:299]: I add a check for token empty like it had before. But I think it's ok if CompileTokens() expects valid utf-8, we can check a string before sending it to CompileTokens().

**Comment by melex750** [comment]
Made changes suggested in review and resolved conflicts with ```dev```.

File: CBot/src/CBot/CBotToken.h [hunk 6]
```diff
@@ -251,4 +283,16 @@ extern bool IsOfTypeList(CBotToken* &p, int type1, ...);
...
+    size_t          currentLineIndex{0};
+    size_t          currentColumnIndex{0};
```
[the remark this answers is not in the document]
Comment (melex750) [reply CBot/src/CBot/CBotToken.h:296]: Maybe something like ```struct CBotCodeStats``` with line count, checksum, etc.

**Review by hexagonrecursion — APPROVED** [review]
Looks good to me

**Pull request merged**